common/core/desktop/src/kmx/kmx_processor.cpp

```diff
@@ -75,43 +103,99 @@ kmx_processor::update_option(
   return option(scope, key, value);
 }
 
+bool kmx_processor::queue_action(
+  km_kbp_action_item const * action_item
+) {
+  DebugLog("Action type is [%d].\n", action_item->type);
+  switch (action_item->type) {
+  case KM_KBP_IT_END:
+    // error should not queue empty item
+    return false;
+    break;
+  case KM_KBP_IT_CHAR:
+   _kmx.GetActions()->QueueAction(QIT_CHAR, action_item->character);
+    break;
+  case KM_KBP_IT_MARKER:
+   _kmx.GetActions()->QueueAction(QIT_DEADKEY, (KMX_DWORD)action_item->marker);
+    break;
+  case KM_KBP_IT_ALERT:
+    _kmx.GetActions()->QueueAction(QIT_BELL, 0);
+    break;
+  case KM_KBP_IT_BACK:
+  {
+    // If the context is already empty, we want to emit the backspace for application to use
+    PKMX_WCHAR p_last_item_context = _kmx.GetContext()->Buf(1);
+
+    if ((!p_last_item_context || *p_last_item_context == 0) ||
+      (action_item->backspace.expected_type == KM_KBP_BT_UNKNOWN)) {
+      _kmx.GetActions()->QueueAction(QIT_INVALIDATECONTEXT, 0);
+      _kmx.GetActions()->QueueAction(QIT_BACK, BK_DEFAULT);
+    } else if (action_item->backspace.expected_type == KM_KBP_BT_MARKER) {
+      _kmx.GetActions()->QueueAction(QIT_BACK, BK_DEADKEY);
+    } else /* KM_KBP_BT_CHAR, KM_KBP_BT_UNKNOWN */ {
+      _kmx.GetActions()->QueueAction(QIT_BACK, BK_DEFAULT);
+    }
+    break;
+  }
+  case KM_KBP_IT_PERSIST_OPT:
+  case KM_KBP_IT_EMIT_KEYSTROKE:
+  case KM_KBP_IT_CAPSLOCK:
+    // Not implemented TODO log message?
+    return false;
+    break;
+  case KM_KBP_IT_INVALIDATE_CONTEXT:
+    _kmx.GetActions()->QueueAction(QIT_INVALIDATECONTEXT, 0);
+    break;
+  }
+  return true;
+}
+
 km_kbp_status
 kmx_processor::process_event(
   km_kbp_state *state,
   km_kbp_virtual_key vk,
   uint16_t modifier_state,
   uint8_t is_key_down
 ) {
-  // Construct a context buffer from the items
-
-  std::u16string ctxt;
-  auto cp = state->context();
-  for (auto c = cp.begin(); c != cp.end(); c++) {
-    switch (c->type) {
-    case KM_KBP_CT_CHAR:
-      if (Uni_IsSMP(c->character)) {
-        ctxt += Uni_UTF32ToSurrogate1(c->character);
-        ctxt += Uni_UTF32ToSurrogate2(c->character);
-      } else {
-        ctxt += (km_kbp_cp)c->character;
+  // If the Virtual Key is VK_SPACE and the internal kmx processor has actions
+  // then process that and return. Why - because these actions must have been added externally
+  // via the queue_action method. The IMX extension requests processing of action queue
```

Comment:
```suggestion
  // via the queue_action method. The client can request processing of action queue
```

We should ensure that this little nuance is documented in the keyboardprocessor.h in `queue_action` and `process_event` comments -- it's an implementation detail that should be made consistent in each of the processors.

---

Alternatively... we could add a tiny little function `km_kbp_process_queued_actions(state)` which avoids this, and refactor `kmx_processor::process_event` and the new `kmx_processor::process_queued_actions` to both call `kmx_processor::internal_process_queued_actions` accordingly? A little bit more work but a whole lot prettier I think. Something like:

```
kmx_processor::process_queued_actions() {
  state->actions().clear();
  return internal_process_queued_actions();
}
```

Reply:
Oh yes this is great. After our initial discussion I saw didn't need to actually synthesize the keystoke using sendinput or whatever -- it would have caused more issues.  I could just send VK_SPACE straight into process_event I didn't make the next logical step that you have here. This is much nicer for the next developer reading the code.